### PR TITLE
chore: update the custom searcher docs [MLG-447]

### PR DESCRIPTION
**Pull request by tayritenour**
## Description

Add a bunch of docstrings and examples to clarify how one should build Custom Searchers

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/searcher/_search_method.py [hunk 1]
```diff
@@ -229,9 +299,43 @@ def on_validation_completed(
...
+
+            def on_validation_completed(
+                self,
+                searcher_state: SearcherState,
+                request_id: uuid.UUID,
+                metric: Any,
+                train_length: int
+            ) -> List[Operation]:
+                MAX_TRAIN_LENGTH = 10  # Best defined in __init__
+                if not searcher_state.experiment_completed \\
```
Comment (garrett361) [inline harness/determined/searcher/_search_method.py:318]: I don't think it's possible to reach this if-statment if either of `searcher_state.experiment_completed` or `searcher_state.experiment_failed` are `True`, right?

File: harness/determined/searcher/_search_method.py [hunk 2]
```diff
@@ -240,15 +344,60 @@ def on_trial_closed(
...
+                    }
+                    return [
+                        searcher.Create(
+                            request_id=uuid.uuid4(),
+                            hparams=hparams,
+                            checkpoint=None,
+                        )
+                    ]
+                if searcher_state.trials_closed >= MAX_NUM_TRIALS:
+                    return [searcher.Shutdown(failure=searcher_state.experiment_failed)]
```
Comment (garrett361) [inline harness/determined/searcher/_search_method.py:369]: Similar question to the above here: is it actually possible for `searcher_state.experment_failed` to be `True` when reaching this code? Didn't think so.

File: harness/determined/searcher/_search_method.py [hunk 3]
```diff
@@ -261,6 +410,39 @@ def on_trial_exited_early(
...
+
+        .. code:: python
+
+            def on_trial_exited_early(
+                self,
+                searcher_state: SearcherState,
+                request_id: uuid.UUID,
+                exited_reason: ExitedReason,
+            ) -> List[Operation]:
+                MAX_FAILURES = 10  # Best defined in __init__
```
Comment (garrett361) [inline harness/determined/searcher/_search_method.py:424]: Personal preference, but the various globals like `MAX_FAILURES` seem a little unexpected. Would instead expect simple attributes: `self.max_failures` and similar.

File: harness/determined/searcher/_search_method.py [hunk 4]
```diff
@@ -261,6 +410,39 @@ def on_trial_exited_early(
...
+                if exited_reason == searcher.ExitedReason.USER_CANCELED:
+                    return [searcher.Shutdown(cancel=True)]
```
Comment (garrett361) [inline harness/determined/searcher/_search_method.py:426]: Didn't remember we had the `cancel` arg for `Shutdown`, nice

File: harness/determined/searcher/_search_method.py [hunk 5]
```diff
@@ -261,6 +410,39 @@ def on_trial_exited_early(
...
+                if exited_reason == searcher.ExitedReason.INVALID_HP:
+                    return [searcher.Shutdown(failure=True)]
+                if searcher_state.failures >= MAX_FAILURES:
+                    return [searcher.Shutdown(failure=True)]
+                return []
+
+        .. note::
+
+            The trial has already been internally closed when this callback is run.
+            You do not need to explicitly issue a :class:`~Close` operation
```
Comment (garrett361) [inline harness/determined/searcher/_search_method.py:436]: Thank you for documenting this!

**Review by garrett361 — APPROVED** [review]
Looks good to me, just small comments.

File: harness/determined/searcher/_search_method.py (hunk 3 above)
Comment (tayritenour) [reply harness/determined/searcher/_search_method.py]: Added

File: harness/determined/searcher/_search_method.py (hunk 2 above)
Comment (tayritenour) [reply harness/determined/searcher/_search_method.py]: Added

**Comment by tayritenour** [comment]
Added all those changes. Thank you!

**Pull request merged**